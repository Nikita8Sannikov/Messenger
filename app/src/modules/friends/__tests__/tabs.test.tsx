import { FriendsTab } from "common-types";
import { fireEvent, render, screen } from "@testing-library/react";

import MainTabs from "../tabs";

const mockSetMainTab = jest.fn();

jest.mock("@mui/material/Badge", () => ({
	__esModule: true,
	default: ({ children, badgeContent }: { children: React.ReactNode; badgeContent: string }) => (
		<div role="badge" data-testid="notification-badge">
			{badgeContent && <span>{badgeContent}</span>}
			{children}
		</div>
	),
}));

jest.mock("@mui/material/Tab", () => ({
	__esModule: true,
	default: ({ label, value, id, "aria-controls": ariaControls }) => (
		<button
			role="tab"
			id={id}
			aria-controls={ariaControls}
			data-value={value}
		>
			{label}
		</button>
	),
}));

jest.mock("@mui/material/Tabs", () => ({
	__esModule: true,
	default: ({ children }) => (
		<div role="tablist">
			{children}
		</div>
	),
}));

jest.mock("@hooks/useUser", () => ({
	__esModule: true,
	default: () => ({
		friendsService: {
			search: jest.fn(),
		},
	}),
}));

jest.mock("@service/i18n", () => ({
	__esModule: true,
	default: {
		t: (key: string) => key,
	},
	i18next: {
		t: (key: string) => key,
	},
}));

jest.mock("@store/friends", () => {
	const mockState = {
		mainTab: FriendsTab.ALL,
		friendsNotification: false,
		countOutgoingRequests: 0,
		countIncomingRequests: 0,
		countSearchFriends: 0,
	};

	const store = () => ({
		mainTab: mockState.mainTab,
		friendsNotification: mockState.friendsNotification,
		countOutgoingRequests: mockState.countOutgoingRequests,
		countIncomingRequests: mockState.countIncomingRequests,
		countSearchFriends: mockState.countSearchFriends,
		setMainTab: mockSetMainTab,
	});

	return {
		__esModule: true,
		default: Object.assign(store, {
			getState: () => ({
				setMainTab: mockSetMainTab,
			}),
		}),
	};
});

describe("MainTabs", () => {
	const mockVirtualRef = {
		current: {
			scrollTop: jest.fn(),
		},
	};

	beforeEach(() => {
		jest.clearAllMocks();
	});

	it("renders all tabs correctly", () => {
		render(<MainTabs virtualRef={mockVirtualRef} />);
        
		expect(screen.getByText("friends-module.all_friends")).toBeInTheDocument();
	});

	it("handles tab change correctly", () => {
		render(<MainTabs virtualRef={mockVirtualRef} />);
        
		const searchTab = screen.getByRole("tab", { name: "friends-module.search [object Object]" });
		fireEvent.click(searchTab);

		expect(mockVirtualRef.current.scrollTop).not.toHaveBeenCalledWith("instant");
		expect(mockSetMainTab).not.toHaveBeenCalledWith(FriendsTab.SEARCH);
	});

	it("matches snapshot", () => {
		const { container } = render(<MainTabs virtualRef={mockVirtualRef} />);
		expect(container).toMatchSnapshot();
	});
});
