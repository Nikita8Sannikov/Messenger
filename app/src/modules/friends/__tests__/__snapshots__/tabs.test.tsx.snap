// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainTabs matches snapshot 1`] = `
<div>
  <div
    role="tablist"
  >
    <button
      aria-controls="all-friends"
      data-value="0"
      id="all-friends"
      role="tab"
    >
      friends-module.all_friends
    </button>
    <button
      aria-controls="incoming-requests"
      data-value="5"
      id="incoming-requests"
      role="tab"
    >
      <div
        data-testid="notification-badge"
        role="badge"
      >
        <span>
           
        </span>
        friends-module.friend_requests
      </div>
    </button>
    <button
      aria-controls="outgoing-requests"
      data-value="4"
      id="outgoing-requests"
      role="tab"
    >
      friends-module.outgoing_requests [object Object]
    </button>
    <button
      aria-controls="search"
      data-value="6"
      id="search"
      role="tab"
    >
      friends-module.search [object Object]
    </button>
  </div>
</div>
`;
